src/execution/mod.rs

```diff
@@ -687,6 +690,11 @@ pub(crate) fn run_main(
     epoch_interval: Option<EpochInterval>,
     recovery_config: Option<Py<RecoveryConfig>>,
 ) -> PyResult<()> {
+    if is_in_bytewax_run(py)? {
```

Comment:
Yes, this was an idea I had yesterday to keep both the old execution methods and the new one, but I don't like it either. I opened this draft PR to get feedback on this, so thanks.

I did this to make the changes a little bit less breaking, but if we are ok on completely redefining the execution model things might get easier.

> Our core dataflow definition is always defined by a "dataflow builder" in a standalone Python file, like your existing get_flow function idea.

Ok, I like this since it avoids the double meaning of the "dataflow-name" argument, which is not good from an api point of view. So we require users to define a getter function.

> Our lowest level entry points like run_main and cluster_main don't know anything about modules, they use fully instantiated Dataflow objects and are maybe considered "private"? They are not meant to be called by users of the library. If people want to do freaky things with Jupyter notebooks, fine, but I think from here on out I'm going to consider that using "private interfaces".

So, I approached this trying to make it easier for existing users to upgrade, but it ended up confusing.
I did it this way because executing any of the running functions during the import leads to a non obvious behaviour: if you try to run with `bytewax.run` and recovery and multiple processes, you see the dataflow running, but it's not running with the config you want. But I agree that just ignoring the code is possibly even more confusing.
Rethinking this, we could move `run_main` (maybe renamed to `run_dataflow`) to `bytewax.testing`   so that it's clear it shouldn't be used in production, and keep the trick to detect if the script was ran with `bytewax.run` and, instead of ignoring the code, just crash the process asking to remove the call?

`cluster_main` still has its reasons for existing: if you want to run processes on different machines both the old and the new `spawn_cluster` wouldn't allow this. And if we remove the "kafka" recovery option from the script, this is the only entrypoint that would allow using it.

I think we either expand the `run` script to allow everything possible with `cluster_main`, and then remove `cluster_main` from the public api completely, or we specify that the `bytewax.run` script is only good for single machine dataflows, but we keep `cluster_main` public and documented for advanced cases.

Currently I'm leaning towards having less entrypoints to avoid confusion, so just `bytewax.testing.run_dataflow` and `bytewax.run`, but it means we have to add some more config options to the script.

> spawn_cluster sits on top of this and does instantiate dataflows from modules and uses the get_flow function and does the process coordination. We can use this in unit tests sometimes (writing those specific dataflows-under-test in a separate test file to be imported), but it isn't really public. (We can also still use run_main and cluster_main in tests too.)

Ok, this means we restrict multiprocess dataflow tests to those specific tests, but I don't think that's bad.

> On to of all this is the run module and python -m bytewax.run. This is the only "user facing" entry point for running dataflows. It looks at the command line flags and decides whether to use run_main, spawn_cluster, or cluster_main internally based on command line flags or env vars.

Which is more or less what we have here. With the latest commits from yesterday I changed things, and now `spawn_cluster` is only using `cluster_main`, but it previously used `run_main` by default and only went for `cluster_main` if specific parameters were passed. I changed it to make the trick of ignoring the code work in some special cases, but if we go for the crashing approach the problem should not be there.
 
> You've thought about this more than me. Perhaps there's some gotchas surrounding doing dynamic imports that makes this difficult.

There are, but the things you suggested mostly help with those, so it should be ok.

I'm going to work on the suggestions and update the PR so we can have a second round of review. I know we want this for the next release, but I don't want to rush this.